### PR TITLE
Persist doc only when no error happens during parsing

### LINKED ISSUE #489: "Document "{document.name}" not added to database, because of parse error" but being added
## Description of the bug

When an error happens during parsing, an error message: "Document "{document.name}" not added to database, because of parse error" appears, but the `Document` object itself is added to database before the parser error happens.

## To Reproduce
1. Use the latest commit (1d6771befb95f4ae94f308899633294a003dcfd6)
2. Let a parser fail
3. Execute `session.query(Document).count()` to see how many `Documents` are added to the database.

## Expected behavior

Either one of these behaviors
1. The `Document` is not added.
2. The error message gets corrected like "Document is added but not parsed".

## Error Logs/Screenshots

"Document "{document.name}" not added to database, because of parse error"

## Environment (please complete the following information)
 - Fonduer Version: 0.8.3-dev (1d6771befb95f4ae94f308899633294a003dcfd6)

## Additional context

I think this is a regression caused by b1b8d24e1133e2229c5c02c6bdcf9eab36301a55, where `doc: Document` is saved to the database by `session.merge(doc, load=True)`, so this could happen on v0.8.0, v0.8.1, v0.8.2 too.

### PR DESCRIPTION
**Pull request by HiromuHota**
## Description of the problems or issues

**Is your pull request related to a problem? Please describe.**

See #489.

**Does your pull request fix any issue.**

Fix #489.

## Description of the proposed changes

Persist doc only when no error happens during parsing.

## Test plan

Added a test that demonstrates #489.

## Checklist

* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] I have updated the CHANGELOG.rst accordingly.

### CODE REVIEW

**Comment by HiromuHota** [comment]
PyTorch 1.6.0 has been released just two days ago.
Mypy complains that "unused 'type: ignore' comment" at `torch.__version__`.
I couldn't track down which change at PyTorch v1.6.0 affects this, but removing that ignore comment satisfies Mypy.

**Comment by HiromuHota** [comment]
Please squash it when merging.

File: src/fonduer/parser/parser.py [hunk 1]
```diff
@@ -141,6 +141,11 @@ def apply(  # type: ignore
             progress_bar=progress_bar,
         )
 
+    def _add(self, doc: Union[Document, None]) -> None:
```
Comment (SenWu) [inline src/fonduer/parser/parser.py:144]: Do we use this function?

Comment (HiromuHota) [reply src/fonduer/parser/parser.py:144]: `Parser._add` is called in the main process after getting a result from `ParserUDF.apply` at
https://github.com/HazyResearch/fonduer/blob/5909011e9bdc52825e480f941a0dff4737002b8e/src/fonduer/utils/udf.py#L143-L150

The `UDF._add` has been meant to add features/labels to the database, more specifically by 

https://github.com/HazyResearch/fonduer/blob/5909011e9bdc52825e480f941a0dff4737002b8e/src/fonduer/features/featurizer.py#L248-L252

and 

https://github.com/HazyResearch/fonduer/blob/5909011e9bdc52825e480f941a0dff4737002b8e/src/fonduer/supervision/labeler.py#L309-L311

.
So only `Featurizer._add` and `Labeler._add` have been implemented, but I realized that `Parser._add` can be implemented to store the "transient" `Document` to database.
We don't need `_add` for `MentionExtractor` and `CandidateExtractor` as extracted mentions and candidates will be persisted without explicitly adding them to the session.

**Pull request merged**